src/agent/block_store_services/block_store_fs.js

```diff
@@ -158,24 +162,47 @@ class BlockStoreFs extends BlockStoreBase {
                 }
             }
 
-            const { data, stat } = await nb_native().fs.readFile(fs_context, block_path, { read_xattr: true });
-
-            // read md from xattr
-            let block_md_from_fs = try_parse_block_md(stat.xattr[config.BLOCK_STORE_FS_XATTR_BLOCK_MD]);
+            let data;
+            let block_md_from_fs;
 
-            // if not able to parse md from xattr, fallback to reading .meta file content which was the old model.
-            if (!block_md_from_fs) {
+            if (this.xattr_fallback) {
+                const meta_path = this._get_block_meta_path(block_md.id);
+                const [block, meta] = await Promise.all([
+                    nb_native().fs.readFile(fs_context, block_path),
+                    nb_native().fs.readFile(fs_context, meta_path),
+                ]);
+
+                data = block.data;
+                block_md_from_fs = try_parse_block_md(meta.data.toString());
+            } else {
                 try {
-                    const meta_path = this._get_block_meta_path(block_md.id);
-                    const { data: meta_data } = await nb_native().fs.readFile(fs_context, meta_path);
-                    block_md_from_fs = try_parse_block_md(meta_data);
-                } catch (err) {
-                    // noop
+                    const { data: block_data, stat } = await nb_native().fs.readFile(fs_context, block_path, { read_xattr: true });
+
+                    data = block_data;
+                    block_md_from_fs = try_parse_block_md(stat.xattr[config.BLOCK_STORE_FS_XATTR_BLOCK_MD]);
+
+                    // if not able to parse md from xattr, fallback to reading .meta file content which was the old model.
+                    if (!block_md_from_fs) {
+                        try {
+                            const meta_path = this._get_block_meta_path(block_md.id);
+                            const { data: meta_data } = await nb_native().fs.readFile(fs_context, meta_path);
+                            block_md_from_fs = try_parse_block_md(meta_data);
+                        } catch (err) {
+                            // noop
+                        }
+                    }
+                } catch (error) {
+                    if (error.code === 'ENOTSUP') {
```

Comment:
we should check that xattr_fallback is not already set so that we don't get into an infinite loop in case of any other kind of ENOTSUP case -

```suggestion
                    if (error.code === 'ENOTSUP' && !this.xattr_fallback) {
```

Reply:
That would be redundant check though, no? This catch will catch exceptions from the else branch will be chosen only when `this.xattr_fallback === false`. We exit the loop because in the recursive call `this.xattr_fallback = true` which means we won't enter this catch.

Am I missing something here?